### PR TITLE
builder: implement support for custom payload/extension data types

### PR DESCRIPTION
**Pull request by ystreet**
The actual writing for these custom types is implementable by implementing a RtpPacketWriter trait.

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -254,93 +217,347 @@ impl<'a> RtpPacketBuilder<'a> {
...
-        let size = self.calculate_size()?;
-        if size > buf.len() {
-            return Err(RtpWriteError::OutputTooSmall(size));
+        writer.finish()
+    }
+
+    /// Write the packet using writer.
+    pub fn write<O>(
+        &self,
+        writer: &mut dyn RtpPacketWriter<Payload = P, Extension = E, Output = O>,
```
Comment (sdroege) [inline src/builder.rs]: Why `dyn` instead of `impl`?

File: src/builder.rs [hunk 2]
```diff
@@ -254,93 +217,347 @@ impl<'a> RtpPacketBuilder<'a> {
...
+    fn max_size(&self) -> Option<usize> {
+        None
+    }
+
+    /// Provides data to append to the output.  May be called multiple times per packet.
+    fn push(&mut self, data: &[u8]);
+
+    /// Provides the extension data to add to the output.  The extension should be written as-is
+    /// without any transformations.
+    fn push_extension(&mut self, extension_data: &Self::Extension);
```
Comment (sdroege) [inline src/builder.rs:340]: I think this is going to help me a lot for implementation extension support in the (de)payloader base classes.

File: src/builder.rs [hunk 3]
```diff
@@ -254,93 +217,347 @@ impl<'a> RtpPacketBuilder<'a> {
...
+
+    fn padding(&mut self, size: u8) {
+        self.padding = Some(size);
+    }
+
+    fn finish(&mut self) -> Self::Output {
+        let mut ret = vec![];
+        if let Some(padding) = self.padding.take() {
+            self.output
+                .extend(std::iter::repeat(0).take(padding as usize - 1));
```
Comment (sdroege) [inline src/builder.rs]:
```suggestion
                .resize(self.output.len() + padding as usize - 1, 0);
```

**Review by sdroege — COMMENTED** [review]
This looks very good otherwise :)

File: src/builder.rs (hunk 1 above)
Comment (ystreet) [reply src/builder.rs]: No reason :)

Comment (sdroege) [reply src/builder.rs]: If you have `impl` then you can also use `dyn` in it, the other way around not :) Leaves the choice to the caller.

**Pull request merged**